File: rust/src/block_scalar.rs
//! Core §6.1.5 literal block scalars (`|`). Shared by the top-level key
//! path (`core.rs`) and the nested block path (`block.rs`): §6.1.5 defines
//! block scalar extent generically ("indentation strictly greater than the
//! indentation of the key that introduced the scalar"), with no top-level
//! restriction, so both call sites must agree byte for byte. Mirrors
//! `js/src/block-scalar.ts`.

use crate::errors::LimaError;
use crate::scalars::check_string_limit;

/// Leading U+0020 count. "Indentation" in §6.1.5 is spaces, from column 0.
fn leading_spaces(line: &str) -> usize {
    line.bytes().take_while(|&b| b == b' ').count()
}

/// Index right after the last non-space byte.
fn trailing_space_end(line: &str) -> usize {
    let b = line.as_bytes();
    let mut end = b.len();
    while end > 0 && b[end - 1] == b' ' {
        end -= 1;
    }
    end
}

/// Builds a `|` block scalar value.
///
/// `body_lines` are the physical source lines *after* the `|` line, each in
/// its original column-0 form (trailing spaces are stripped here).
/// `key_indent` is the indentation of the introducing key (0 at the top
/// level). `pipe_line` is the 1-based line of the `|` line, so
/// `body_lines[i]` is at `pipe_line + 1 + i`.
///
/// Returns the joined value, the raw body (original column-0 lines joined
/// with `\n`, from which a token's physical `(line, offset)` is read — §2.4,
/// never from the `^^`-merged `joined` string), and how many of `body_lines`
/// the scalar consumed.
pub(crate) fn build_block_scalar(
    body_lines: &[&str],
    key_indent: usize,
    pipe_line: u32,
) -> Result<(String, String, usize), LimaError> {
    // Extent (§6.1.5): a line belongs to the scalar iff its indentation is
    // strictly greater than the introducing key's. Empty lines between
    // content lines belong regardless; the first non-empty line dedented to
    // the key's column or less ends the scalar, `#` lines included.
    let mut consumed = body_lines.len();
    for (i, line) in body_lines.iter().enumerate() {
        let ind = leading_spaces(line);
        if ind == line.len() {
            continue; // empty
        }
        if ind <= key_indent {
            consumed = i;
            break;
        }
    }
    let lines = &body_lines[..consumed];

    // Content indentation (§6.1.5): the smallest number of leading spaces
    // among all non-empty content lines, measured from column 0, removed
    // uniformly from every line.
    let mut min_indent = usize::MAX;
    for line in lines {
        let ind = leading_spaces(line);
        if ind == line.len() {
            continue; // empty lines do not participate
        }
        min_indent = min_indent.min(ind);
    }
    let trim_amt = if min_indent == usize::MAX {
        0
    } else {
        min_indent
    };

    let mut merged: Vec<String> = Vec::new();
    for line in lines {
        let b = line.as_bytes();
        let mut start = trim_amt.min(line.len());
        let is_continuation = b.get(start) == Some(&b'^') && b.get(start + 1) == Some(&b'^'); // ^^
        if is_continuation {
            start += 2;
        }
        let mut end = trailing_space_end(line);
        if end < start {
            end = start;
        }
        let content = &line[start..end];
        if is_continuation && !merged.is_empty() {
            if !content.is_empty() {
                let last = merged.last_mut().unwrap();
                last.push(' ');
                last.push_str(content);
            }
        } else {
            merged.push(content.to_string());
        }
    }

    // Trailing content (§6.1.5): all trailing empty lines are stripped.
    while merged.last().is_some_and(String::is_empty) {
        merged.pop();
    }

    let joined = merged.join("\n");
    check_string_limit(&joined, pipe_line)?;
    Ok((joined, lines.join("\n"), consumed))
}

File: js/src/block-scalar.ts
/**
 * Core §6.1.5 literal block scalars (`|`). Shared by the top-level key path
 * (`core.ts`) and the nested block path (`block.ts`): §6.1.5 defines block
 * scalar extent generically ("indentation strictly greater than the
 * indentation of the key that introduced the scalar"), with no top-level
 * restriction, so both call sites must agree byte for byte.
 *
 * Generic over the output representation via `ValueBuilder` (see
 * `builder.ts`), like the rest of the Core grammar.
 */

import { LString } from './value.js'
import { type ParseContext, checkScalarLimit } from './normalize.js'
import type { ValueBuilder } from './builder.js'

/** Leading U+0020 count. "Indentation" in §6.1.5 is spaces, measured from column 0. */
export const leadingSpaces = (line: string): number => {
	let i = 0
	while (i < line.length && line.charCodeAt(i) === 32) i++
	return i
}

/** Index right after the last non-space character — the `trimEnd()` boundary, without allocating. */
export const trailingSpaceEnd = (line: string): number => {
	let i = line.length
	while (i > 0 && line.charCodeAt(i - 1) === 32) i--
	return i
}

/**
 * Builds a `|` block scalar value.
 *
 * `bodyLines` are the physical source lines *after* the `|` line, each in
 * its original column-0 form (trailing whitespace is stripped here).
 * `keyIndent` is the indentation of the key that introduced the scalar
 * (0 at the top level). `pipeLine` is the 1-based line number of the `|`
 * line, so `bodyLines[i]` is at `pipeLine + 1 + i`.
 *
 * Returns the joined value plus how many of `bodyLines` the scalar
 * consumed — a caller iterating physical lines resumes past that; a caller
 * that already sliced the range to the scalar can ignore it.
 */
export const buildBlockScalar = <V, M>(
	bodyLines: string[], keyIndent: number, pipeLine: number,
	ctx: ParseContext, builder: ValueBuilder<V, M>,
): { value: V; consumed: number } => {
	// Extent (§6.1.5): a line belongs to the scalar iff its indentation is
	// strictly greater than the introducing key's. Empty lines between
	// content lines belong regardless of indentation; the first non-empty
	// line dedented to the key's column or less ends the scalar, `#` lines
	// included.
	let consumed = bodyLines.length
	for (let i = 0; i < bodyLines.length; i++) {
		const bl = bodyLines[i]
		const ind = leadingSpaces(bl)
		if (ind === bl.length) continue // empty (blank or whitespace-only)
		if (ind <= keyIndent) { consumed = i; break }
	}
	const lines = consumed === bodyLines.length ? bodyLines : bodyLines.slice(0, consumed)

	// Content indentation (§6.1.5): the smallest number of leading spaces
	// among all non-empty content lines, measured from column 0, removed
	// uniformly from every line.
	let minIndent = Infinity
	for (const l of lines) {
		const ind = leadingSpaces(l)
		if (ind === l.length) continue // empty lines do not participate
		if (ind < minIndent) minIndent = ind
	}
	const trimAmt = isFinite(minIndent) ? minIndent : 0

	// dedent + `^^` continuation strip + trailing-space trim collapse into a
	// single slice() per line. leadingSpaces/trailingSpaceEnd avoid the
	// temporary strings trim()/trimStart()/trimEnd() would allocate.
	const merged: string[] = []
	for (let i = 0; i < lines.length; i++) {
		const l = lines[i]
		const lineLen = l.length
		let start = trimAmt < lineLen ? trimAmt : lineLen
		const isContinuation = l.charCodeAt(start) === 94 && l.charCodeAt(start + 1) === 94 // ^^
		if (isContinuation) start += 2
		let end = trailingSpaceEnd(l)
		if (end < start) end = start
		const content = l.slice(start, end)
		if (isContinuation && merged.length > 0) {
			if (content) merged[merged.length - 1] += ' ' + content
		} else {
			merged.push(content)
		}
	}

	// Trailing content (§6.1.5): all trailing empty lines are stripped.
	while (merged.length > 0 && merged[merged.length - 1] === '') merged.pop()

	const joined = merged.join('\n')
	checkScalarLimit(LString(joined), pipeLine)
	// A token's physical `(line, offset)` is read from the original body
	// lines (column-0 form), not reconstructed from the dedented/`^^`-merged
	// `joined` string — References 2.0 §2.4. `lines` are at physical lines
	// `pipeLine + 1 .. pipeLine + consumed` (0-based indices `pipeLine ..`).
	return {
		value: builder.string(joined, pipeLine + 1, false, {
			raw: lines.join('\n'), line: pipeLine + 1, col: 0,
			tabAdjust: ctx.tabAdjust?.slice(pipeLine, pipeLine + consumed),
		}),
		consumed,
	}
}
